### PR TITLE
fixes to allow mac builds to work smoothly

### PR DESCRIPTION
**Pull request by ritvikrao**
I made a couple of changes to allow the charm4py build on macs to work smoothly. Note that this only solves the issues that arise after a successful charm++ install has happened; there's an ongoing pr in the charm++ repo to fix charm4py installation issues there.

The first change adds a flag in setup.py to ignore implicit function declaration errors that show up with newer versions of Apple Clang (15.0.0 and up). The second change modifies charm4py/charmlib/ctypes_py to specify that the charm++ library file created by mac builds is libcharm.dylib, not libcharm.so.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -51,6 +51,7 @@ def get_build_triple(build_mpi):
     os.environ['ARCHFLAGS'] = f'-arch {machine}'
     libcharm_filename = 'libcharm.dylib'
     charmrun_filename = 'charmrun'
+    os.environ['CPPFLAGS'] = f'-Wno-error=implicit-function-declaration'
```
Comment (matthiasdiener) [inline setup.py:54]: Is this change necessary due to the missing whole-archive option in https://github.com/UIUC-PPL/charm/pull/3797 ?

**Comment by ritvikrao** [comment]
I think this should be merged now, with the fix made here: https://github.com/UIUC-PPL/charm/pull/3797

File: setup.py (hunk 1 above)
Comment (ritvikrao) [reply setup.py:54]: Yes, it's needed for the newer AppleClang compiler, it's not related to the whole archive issue

Comment (matthiasdiener) [reply setup.py:54]: Can you add a comment to the file on why it is necessary?

Comment (ritvikrao) [reply setup.py]: done

Comment (matthiasdiener) [reply setup.py]: Can you clarify more - what's the error message(s)? Also, are we certain that just disabling the warning is the right thing to do, and not just masking another issue?

Comment (ritvikrao) [reply setup.py]: The errors are these:

```
charm4py/charmlib/charmlib_cython.c:12983:3: error: call to undeclared function 'CkExtContributeToChare'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  CkExtContributeToChare((&__pyx_v_contributeInfo->internal), ((int)__pyx_t_3), __pyx_v_objPtr);
  ^
charm4py/charmlib/charmlib_cython.c:13127:3: error: call to undeclared function 'CkExtContributeToGroup'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  CkExtContributeToGroup((&__pyx_v_contributeInfo->internal), __pyx_v_gid, __pyx_v_elemIdx);
  ^
charm4py/charmlib/charmlib_cython.c:13316:3: error: call to undeclared function 'CkExtContributeToArray'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  CkExtContributeToArray((&__pyx_v_contributeInfo->internal), __pyx_v_aid, __pyx_v_8charm4py_8charmlib_15charmlib_cython_c_index, __pyx_v_ndims);
  ^
charm4py/charmlib/charmlib_cython.c:13483:3: error: call to undeclared function 'CkExtContributeToSection'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  CkExtContributeToSection((&__pyx_v_contributeInfo->internal), __pyx_t_2, __pyx_t_3, __pyx_v_rootPE);
  ^
charm4py/charmlib/charmlib_cython.c:14889:3: error: call to undeclared function 'LBTurnInstrumentOn'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  LBTurnInstrumentOn();
  ^
charm4py/charmlib/charmlib_cython.c:14924:3: error: call to undeclared function 'LBTurnInstrumentOff'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  LBTurnInstrumentOff();
```

For clang 15, they turned the implicit function declaration issue from a warning to an error (https://releases.llvm.org/15.0.0/tools/clang/docs/ReleaseNotes.html), and the release notes for clang 15 say that with C2x and later, the error can't be erased with the flag. So eventually this flag option won't work, but for now all the tests are passing with this fix.

Comment (ritvikrao) [reply setup.py]: since these issues are with cythonizing the charmlib_cython.pyx file, it's possible the issues go away with the updates to cython3, but I'm not sure how to go about testing that

Comment (matthiasdiener) [reply setup.py]: These do look fixable - 

> The errors are these:
> 
> ```
> charm4py/charmlib/charmlib_cython.c:12983:3: error: call to undeclared function 'CkExtContributeToChare'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   CkExtContributeToChare((&__pyx_v_contributeInfo->internal), ((int)__pyx_t_3), __pyx_v_objPtr);
>   ^
> charm4py/charmlib/charmlib_cython.c:13127:3: error: call to undeclared function 'CkExtContributeToGroup'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   CkExtContributeToGroup((&__pyx_v_contributeInfo->internal), __pyx_v_gid, __pyx_v_elemIdx);
>   ^
> charm4py/charmlib/charmlib_cython.c:13316:3: error: call to undeclared function 'CkExtContributeToArray'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   CkExtContributeToArray((&__pyx_v_contributeInfo->internal), __pyx_v_aid, __pyx_v_8charm4py_8charmlib_15charmlib_cython_c_index, __pyx_v_ndims);
>   ^
> charm4py/charmlib/charmlib_cython.c:13483:3: error: call to undeclared function 'CkExtContributeToSection'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   CkExtContributeToSection((&__pyx_v_contributeInfo->internal), __pyx_t_2, __pyx_t_3, __pyx_v_rootPE);
>   ^

Can you copy the prototypes from 
https://github.com/UIUC-PPL/charm/blob/4ad4095a34cf22f122e1d995d4acaec84fef2f4f/src/ck-core/ckreduction.C#L1883 here? Later on, we could go ahead and properly export those functions in Charm++.


> charm4py/charmlib/charmlib_cython.c:14889:3: error: call to undeclared function 'LBTurnInstrumentOn'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   LBTurnInstrumentOn();
>   ^
> charm4py/charmlib/charmlib_cython.c:14924:3: error: call to undeclared function 'LBTurnInstrumentOff'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   LBTurnInstrumentOff();
> ```

Can you try including `LBManager.h` for this error?

> For clang 15, they turned the implicit function declaration issue from a warning to an error (https://releases.llvm.org/15.0.0/tools/clang/docs/ReleaseNotes.html), and the release notes for clang 15 say that with C2x and later, the error can't be erased with the flag. So eventually this flag option won't work, but for now all the tests are passing with this fix.

Ok, but it would probably be better longer-term to fix these instead of switching off the warning/error.

Comment (ritvikrao) [reply setup.py]: I'm not sure how to add the prototypes and the LBmanager import, since the errors are from a generated file

Comment (ritvikrao) [reply setup.py]: I also re-ran this script on Delta, and the errors come up as warnings (because the gcc compiler is older there), so this is not a Mac-specific issue.

**Review by matthiasdiener — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (matthiasdiener) [reply setup.py]: > I'm not sure how to add the prototypes and the LBmanager import, since the errors are from a generated file

Maybe similar to what how it is done in https://github.com/UIUC-PPL/charm4py/blob/8d8b881108215869d1c77c18ed476347526f31e4/charm4py/charmlib/ccharm.pxd ?

**Comment by ritvikrao** [comment]
I added changes to the charm4py install documentation to put the current working build procedure first. Additionally, as Sanjay said in the 4/1 Charm4py meeting, any fixes to the implicit function warnings should be part of another pr.

File: docs/install.rst [hunk 2]
```diff
@@ -72,25 +76,24 @@ and run setup::
     to manually build the Charm++ library (see below).
 
 
-Manually building the Charm++ shared library
---------------------------------------------
+pip (Python 3.7 and older)
```
Comment (adityapb) [inline docs/install.rst:79]: `pip` works for newer Python versions too. Just that `pip3 install charm4py`  will build from source instead of getting prebuilt binaries

**Comment by ritvikrao** [comment]
I added an OS call to fix the path in the c_object_store shared object file

**Comment by ZwFink** [comment]
We should check that this also works for Apple Silicon macs... @mayantaylor can you confirm?

**Comment by adityapb** [comment]
@ritvikrao  Can you make these changes on top of the main branch so that my commits from the ray branch don't show up
